### ISSUE TITLE
Pretrained Model Output Permuted

### ISSUE THREAD

**Issue by hendrycks** [CSAILVision/places365#22]
It appears the pre-trained pytorch model _consistently_ predicts a label different from what it should. For example, instead of predicting 2, the model predicts 10; instead of predicting 3, it predicts 100.

Here is some code to reproduce this issue.

```python
import argparse
import torch
import torch.backends.cudnn as cudnn
import torch.nn.functional as F
import torchvision.datasets as dset
import torchvision.transforms as transforms
import torchvision.models as models
from torch.autograd import Variable
import numpy as np

parser = argparse.ArgumentParser(description='Demo of bug',
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
# Optimization options
parser.add_argument('--batch_size', '-b', type=int, default=100, help='Batch size.')
parser.add_argument('--test_bs', type=int, default=100)
# Acceleration
parser.add_argument('--ngpu', type=int, default=1, help='0 = CPU.')
parser.add_argument('--prefetch', type=int, default=5, help='Pre-fetching threads.')
args = parser.parse_args()

mean = [0.485, 0.456, 0.406]
std = [0.229, 0.224, 0.225]
transform = transforms.Compose(
    [transforms.Scale(256), transforms.CenterCrop(224), transforms.ToTensor(), transforms.Normalize(mean, std)])

test_data = dset.ImageFolder(root="/share/data/lang/users/dan/datasets/places365/val", transform=transform)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=args.test_bs, shuffle=False,
                                          num_workers=args.prefetch, pin_memory=True)

net = models.resnet50(num_classes=365)
checkpoint = torch.load('/share/data/lang/users/dan/.torch/models/resnet50_places365.pth')
state_dict = {str.replace(k,'module.',''): v for k,v in checkpoint['state_dict'].items()}
net.load_state_dict(state_dict)
net.eval()

for p in net.parameters():
    p.volatile = True

if args.ngpu > 1:
    net = torch.nn.DataParallel(net, device_ids=list(range(args.ngpu)))
if args.ngpu > 0:
    net.cuda()

np.random.seed(1)
torch.manual_seed(1)
if args.ngpu > 0:
    torch.cuda.manual_seed(1)

cudnn.benchmark = True  # fire on all cylinders

to_np = lambda x: x.data.cpu().numpy()

for batch_idx, (data, target) in enumerate(test_loader):
    data = Variable(data.cuda(), volatile=True)

    output = net(data)
    smax = to_np(F.softmax(output))
    # batch_idx = target since batch size = 100, the size of each val folder
    print(np.argmax(smax, axis=1), batch_idx)

```

Classes 0 and 1 are correctly predicted. After that, 2 -> 10, 3 -> 100, 4 -> 101, 5 -> 102, 6 -> 103, ...

```python
[  0   0   0 170 174   0  18 174 293 170   0 170 293   0 293 174   0   0
 174   0   0   0   0 174 174 293   0   0 169 170   0 293   0   0 293 293
 186   0 174   0 174   0   0 207 293   0 293   0 192   0   0   0   0   0
   0   0   0   0   0 174   0   0   0   0   0 140   0   0   0   0 293 293
   0   0   0   0 293   0 174 293 174 174 293   0 293   0   0   0 293 140
   0 293   0 174   0 293 293 293 174   0] 0
[ 38   1   1 346   1   1  58   1   1   1   1  98   1   1  98   1   1   1
   1   1   1   1  27   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
   1   1 336   1   1   1   1   1 336   1   1   1   1   1   1   1  98   1
   1   1   1   1   1  55   1  98   1   1   1   1   1   1   1   1   1   2
 135  55   1   1   1   1 298   1   1  82] 1
[ 10 113  10  10 278  84  84 109 347  10 347 347 113  66 218  10  10  10
  10  10 347 218  10  10  10  10 347  10 292 229  10  10 347  10  78 347
 347  10  10  10  66  10  10  10 347  10 349  12  10  10  10  10  10  10
  10 283  10  66  10  10  10 218  10  81  10  10  66 347 347  33  10  10
 292  10 292  10 347 307 127  10  10 347 347  66  10 347  10  10  10  10
  10  85 273  10 347  10  10  10  10  10] 2
[100 238 100 176 176 100 246 100 100 176  56 100 100 235 176 244 182 100
 298 100 264 176 176 244 100 177 264 100 210 246 100 176 210 246 100 100
 244 246 100 212 100 244 102 100 100  63 244 238 244 244 202 100 100 100
 176 246 211 177 100 100 100 177 176 246 100  54 176 100 100 100 100 100
 100 246 244  33 100 238 100 100  19 100 176  20 176 100 100 246 244 246
 100 100 176 100 100 100 100 244 100 100] 3
[ 38 210 101 177 210 101 101 102 210 280 210  16 101 102  27  38 101 211
 101 101 101  27 101 210 101 211 211 101  92 101 101 101  27 102 211  38
 101 101 101 101 101 211 101 101 101 210  27 241 102 211  38 102 101 235
 101 101 211 211 102 211 235 101 211 101 101 210 101 211  27 211 101  18
 101 211 101 244 101  27 101 211 101 101  27  38 101 101 101 101 210 101
 210 211  27 101 211 235  38 101 101 101] 4
[102 102 102 102 101  38 211 102 102 211 102 121 211  38 102 102 102 102
 102 219 102 211 102 246  27 121 102 102  27 121 102 102 102  38 102 244
 102 101 102 102 244 120 210 102 102 102 102 211 100 102 102 102 235 212
 101 102  20 210 102 102 102 128 102 101 102 102 102 101 102 102 102 102
 102 211 102 102 210 102 210 101 211 102 102 102 102 102 102 102 102 102
 210 121 102 102 102 248 244  38 101 102] 5
[103  13 308 192 216 218 192  66 103 136 206 136 226 103 103 294 307 245
 255 103 125 103 103 192   8 216 308 216 125 103 143 165 103 340 256 103
  76 127 125 256 339 103 103 247 103 103 127 103 103 136 103 103 103 103
 103 143  13 103 103   7 103 171 103 338 307 103 136 103 103 103 131 103
 340 103 318 127 228 103  41 103  13  71 103 103 192 131 103 143  19 103
 331 166 339  43 103   5 221 223 103 169] 6
```

With these commands I got the validation data suitable for pytorch's ImageFolder.
```python
import os
for i in range(365):
    os.mkdir('./val/' + str(i))
with open('places365_val.txt') as f:
    for line in f:
        line = line.split()
        os.rename('./val/images/' + line[0], "./val/"+str(line[1])+'/'+line[0])
```

**Comment by metalbubble** [comment]
hmmm, is the result consistent with the testing script (https://github.com/CSAILVision/places365/blob/master/test_placesCNN_pytorch.py)? The output of the testing script seems all-right.

**Comment by hendrycks** [comment]
It appears the ImageFolder dataset creator is sorting the classes in a way I did not expect.

`classes = sorted([str(i) for i in range(365)])`
yields the permutation
`['0', '1', '10', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '11', '110', '111', '112' ,...]`
which is what I observe.

Thus this is caused by the ImageFolder dataset creator.
